scipy/interpolate/src/_fitpackmodule.c

```diff
@@ -191,14 +192,16 @@ fitpack_bispev(PyObject *dummy, PyObject *args)
     ny = ap_ty->dimensions[0];
     mx = ap_x->dimensions[0];
     my = ap_y->dimensions[0];
-    mxy = mx*my;
-    if (my != 0 && mxy/my != mx) {
+
+    if (my != 0 && INTPTR_MAX/my < mx) {
         /* Integer overflow */
         PyErr_Format(PyExc_RuntimeError,
                      "Cannot produce output of size %dx%d (size too large)",
                      mx, my);
         goto fail;
     }
+
+    mxy = mx*my;
```

Comment:
Sorry, you still need to cast before doing the multiplication to handle cases where `F_INT` is smaller than `npy_intp`. It's just that we don't need the intermediate unsigned type.
```suggestion
    mxy = (npy_intp)mx*(npy_intp)my;
```

Reply:
@Lechnio did this solution not work?

Reply:
no, cuz it checks if mx*my would extend INTPTR_MAX value, but due to overflow it would never happen.

Current solution with size_t casting is correct, but it would not success on all pipelines, cause they have to be fixed (also some scipy tests). There are just more bugs that have been discovered after this change.

Reply:
>no, cuz it checks if mx*my would extend INTPTR_MAX value

Which is why, as I said, you need to check `INTPTR_MAX/my < mx`.

>Current solution with size_t casting is correct

No it isn't. `size_t` has a greater range than `INTPTR_MAX` so can still have invalid values.

Reply:
width of `size_t` is dynamic and depends on architecture type as INTPTR does please see implementation sources.

Reply:
That's true but on all platforms I'm aware of `sizeof(size_t) == sizeof(intptr_t)` so this has the exact same problems when you were casting to `uintptr_t` before.

Reply:
but `uintptr_t`n requires another include, and  I dont know if that include is kind of platform specific, but it could be (e.g. like windows.h on windows) so that is not good idea unless you provide common include.

if you look in history i used your solution weeks ago and it wont work. pipeline are broken and code behavior is same as none changes were mage. i tested in either on32bit device and 64bit, this changes are not enough. I did not analyze it but maybe compiler changes `INTPTR_MAX/my < mx` to `INTPTR_MAX < mx*my` since multiplication is easier for computing 

method with casting on `size_t` is correct, and works either with -O2 and -O3 flag (O2 is default applied in scipy)

Reply:
>if you look in history i used your solution weeks ago and it wont work.

You've forced pushed so I can't see your commit history. In the only version I saw, your version was broken because you forgot to cast `mx` and `my` to `npy_intp`. Which is the original comment I made in this thread.

>MAYBE compiler changes `INTPTR_MAX/my < mx` to `INTPTR_MAX < mx*my` since multiplication is easier for computing

The compiler is allowed to assume your calculations don't overflow an `int` because signed integer overflow is undefined behaviour. However, it can't transform those into calculation into something which would overflow. No compiler would make that optimization. 

The original test fails because the compiler computes the division and comparison  in 64 bit registers; without overflow, this gives the same result and so is a valid optimization. However, if `mxy` is only 32-bit then it won't catch the overflow. `INTPTR_MAX/my < mx` doesn't suffer from this because the largest value `mxy` can hold is given explicitly instead of relying implicitly on the length of the register.

> method with casting on `size_t` is correct, 

`mxy` can be greater than `INTPTR_MAX` but less than `SIZE_MAX` in which case this overflow check will pass but constructing the array on line 202 will fail:
https://github.com/scipy/scipy/blob/8c7973148c5dea4a28e0e082f8c1f182313b72c1/scipy/interpolate/src/_fitpackmodule.c#L202

`PyArray_SimpleNew` takes a `npy_intp` so simply changing the type of `mxy` doesn't fix anything.

Reply:
`SIZE_MAX` depends on arch type as `size_t` does so there wont be any difference between `INTPTR_MAX` `SIZE_MAX`. I did cast `mx` and `my` inside my test and your solution failed :) and as I checked, minimal code changes with casting on `size_t` works either o 32 bit arm and 64bit amd, also assembly code looks correct, cause overflow of unsigned types is compliant with the ANSI C and C++ standard.

Reply:
>there wont be any difference between `INTPTR_MAX` and `SIZE_MAX`

One is a signed type, the other is unsigned. They are not the same.

Reply:
This code demonstrates the problem:
https://www.ideone.com/Ewrrew#stdin

There are values which overflow signed types but not unsigned types. Your code misses these.

Reply:
>Yes but I do not use them so there is no difference if they are or not.

Don't use what?